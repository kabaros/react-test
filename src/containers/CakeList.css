.cakes-container {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    text-align: center;
    margin-top: 1rem;
}