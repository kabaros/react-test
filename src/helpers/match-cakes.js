
export default (allCakes, cakeName) => {
    const matchCakes = (cake) => cake.title.match(new RegExp(cakeName, 'i'));
    return allCakes.filter(cake => matchCakes(cake, cakeName));   
};