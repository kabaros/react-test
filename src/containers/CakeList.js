import React, { Component } from 'react';
import CakeCard from '../components/CakeCard';
import CakeSearch from '../components/CakeSearch';
import { fetchCakes } from '../helpers/api-helpers';
import matchCakes from '../helpers/match-cakes';

import './CakeList.css';

class CakeList extends Component {
    state = {
        cakes: [],
        searchCriteria: '',
    }

    allCakes = []

    async componentDidMount() {
        const cakes = await fetchCakes();
        this.allCakes = cakes;
        
        this.setState({
            cakes
        });
    }

    searchCakes = (cakeName) => {
        if(!cakeName) {
            this.setState({
                cakes: this.allCakes
            })
        }
        const cakes = matchCakes(this.allCakes, cakeName);

        this.setState({
            cakes
        })
    }

    render() {
        return (
        <div>
            <CakeSearch searchCriteria={this.state.searchCriteria} onSearch={this.searchCakes} />
            <div className="cakes-container">
                {
                    /*
                        I am using the index as a key to suppress errors
                        This is wrong - the api should have provided a unique key to use
                     */
                    this.state.cakes.map((cake, i) => <CakeCard key={i} cake={cake} />)
                }
            </div>
        </div>);
    }
}

export default CakeList;