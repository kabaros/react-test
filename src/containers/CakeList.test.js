import React from 'react';
import CakeList from './CakeList';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import helpers from '../helpers/api-helpers';

jest.mock('../helpers/api-helpers', () => {
    const fetchCakes = jest.fn();
    fetchCakes.mockReturnValue([{
        desc: 'some description',
        img: 'some-image-url',
        title: 'awesome-cake'
    }, {
        desc: 'bad cake description',
        img: 'bad-image-url',
        title: 'bad-cake'
    }]);

    return {
        fetchCakes
    }
});

it('should call the api', () => {
    renderer.create(<CakeList />);
    expect(helpers.fetchCakes).toHaveBeenCalled();
});

it('should render the list of cakes returned by the api', async() => {
  const component = await renderer.create(<CakeList />);
  expect(component).toMatchSnapshot();
});

describe('searching cakes', () => {
    it('should update based on search criteria', async () => {
        const component = await mount(<CakeList />);
        const input = component.find('input');
        input.simulate('change',  {target: {value: 'bad'}});

        const state = component.state();
        expect(state.cakes.length).toEqual(1);
        expect(state.cakes[0].desc).toEqual('bad cake description');
    });

    it('should show all cakes after deleting the search criteria', async () => {
        const component = await mount(<CakeList />);

        // searching for bad cake
        const input = component.find('input');
        input.simulate('change',  {target: {value: 'bad'}});
        expect(component.state().cakes.length).toEqual(1);

        // deleting search criteria
        input.simulate('change',  {target: {value: ''}});
        expect(component.state().cakes.length).toEqual(2);
    });
});
