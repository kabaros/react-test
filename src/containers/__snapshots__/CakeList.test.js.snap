// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the list of cakes returned by the api 1`] = `
<div>
  <div
    className="cake-search"
  >
    <input
      onChange={[Function]}
      placeholder="Search for a cake"
      type="text"
    />
  </div>
  <div
    className="cakes-container"
  >
    <article
      className="cake-card"
    >
      <div
        className="card-header"
      >
        <img
          alt="awesome-cake"
          src={undefined}
        />
      </div>
      <h2>
        awesome-cake
      </h2>
      <p>
        some description
      </p>
    </article>
    <article
      className="cake-card"
    >
      <div
        className="card-header"
      >
        <img
          alt="bad-cake"
          src={undefined}
        />
      </div>
      <h2>
        bad-cake
      </h2>
      <p>
        bad cake description
      </p>
    </article>
  </div>
</div>
`;
